src/arithmetic/samplable.rs

```diff
@@ -0,0 +1,54 @@
+use num_integer::Integer;
+use rand::{rngs::OsRng, RngCore};
+
+use super::traits::{BitManipulation, Converter, One, Samplable, Zero};
+use super::BigInt;
+
+impl Samplable for BigInt {
+    fn sample_below(upper: &Self) -> Self {
+        assert!(*upper > Self::zero());
+
+        let bits = upper.bit_length();
+        loop {
+            let n = Self::sample(bits);
+            if n < *upper {
+                return n;
+            }
+        }
+    }
+
+    fn sample_range(lower: &Self, upper: &Self) -> Self {
+        assert!(upper > lower);
+        lower + Self::sample_below(&(upper - lower))
+    }
+
+    fn strict_sample_range(lower: &Self, upper: &Self) -> Self {
+        assert!(upper > lower);
+        loop {
+            let n = lower + Self::sample_below(&(upper - lower));
+            if n > *lower && n < *upper {
+                return n;
+            }
+        }
+    }
+
+    fn sample(bit_size: usize) -> Self {
```

Comment:
in the below two functions I see deviations from original code. 
can you justify those ? 
I prefer to not make changes at all if possible. sampling randomness is delicate and the function here went through extensive battle testing

Reply:
`sample` function is the same — I've just added corner-case check if bit_size = 0 (sample used to overflow uint in this case).

I changed a lot `strict_sample` function. It used to call `sample` in the loop until random int meets required size in bits. I've just removed loop by using the fact that the most left bit is always 1. But I agree that it's not the case of PR, I will put back old implementation.